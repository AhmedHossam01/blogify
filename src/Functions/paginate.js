const numberPerPages = 6;

export const paginate = (posts, currentPage) => {
  const firstIndex = numberPerPages * currentPage;
  const lastIndex = firstIndex + numberPerPages;
  const currentPosts = posts.slice(firstIndex, lastIndex);
  return currentPosts;
};

export const getNumbersOfPages = posts => {
  const numberOfPages = Math.ceil(posts.length / numberPerPages);
  return numberOfPages;
};

export const getPages = posts => {
  const numberOfPages = getNumbersOfPages(posts);
  let pages = [];
  for (let i = 0; i < numberOfPages; i++) {
    pages.push(i);
  }
  return pages;
};
