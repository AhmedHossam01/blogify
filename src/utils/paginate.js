const numberPerPage = 6;

export const paginate = (posts, currentPage) => {
  const firstIndex = numberPerPage * currentPage;
  const lastIndex = firstIndex + numberPerPage;
  const currentPosts = posts.slice(firstIndex, lastIndex);
  return currentPosts;
};

export const getNumbersOfPages = posts => {
  const numberOfPages = Math.ceil(posts.length / numberPerPage);
  return numberOfPages;
};

export const getPages = posts => {
  const numberOfPages = getNumbersOfPages(posts);
  let pages = [];
  for (let i = 0; i < numberOfPages; i++) {
    pages.push(i);
  }
  return pages;
};
