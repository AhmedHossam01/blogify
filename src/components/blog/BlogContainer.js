import React, { Component } from "react";
import BlogPost from "./BlogPost";
import CardColumns from "react-bootstrap/CardColumns";
import paginate from "../../utils/paginate";

export default class BlogContainer extends Component {
  getPosts = () => {
    return paginate.getItems(
      this.props.posts,
      this.props.currentPage,
      this.props.numberPerPage
    );
  };

  render() {
    const currentPosts = this.getPosts();

    return (
      <CardColumns>
        {currentPosts.map((post, index) => {
          return (
            <BlogPost
              key={index}
              title={post.title}
              body={post.body}
              picture="image.jpg"
            ></BlogPost>
          );
        })}
      </CardColumns>
    );
  }
}
