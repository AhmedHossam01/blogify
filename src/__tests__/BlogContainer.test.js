import React from "react";
import { render } from "react-dom";
import BlogContainer from "../components/blog/BlogContainer";
import { act } from "@testing-library/react";
import paginate from "../utils/paginate";

describe("Blog container use paginate function and renders correctly", () => {
  const fakePosts = [
    {
      userId: 1,
      id: 1,
      title:
        "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
      body:
        "quia et suscipitsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
    },
    {
      userId: 1,
      id: 2,
      title: "qui est esse",
      body:
        "est rerum tempore vitaesequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
    }
  ];

  jest.spyOn(paginate, "getItems");
  const container = document.createElement("div");

  it("calls paginate function correctly", () => {
    act(() => {
      render(
        <BlogContainer
          posts={fakePosts}
          numberPerPage={6}
          currentPage={0}
        ></BlogContainer>,
        container
      );
    });

    // The paginate function is called one time
    expect(paginate.getItems.mock.calls.length).toBe(1);

    // The first argument of the first call of this function was fakePosts
    expect(paginate.getItems.mock.calls[0][0]).toBe(fakePosts);

    // It's clear!
    expect(paginate.getItems).toHaveBeenCalledWith(fakePosts, 0, 6);
    expect(container).toMatchSnapshot();
    paginate.getItems.mockClear();
  });
});
