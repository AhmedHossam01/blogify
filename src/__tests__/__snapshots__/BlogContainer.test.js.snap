// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog container use paginate function and renders correctly calls paginate function correctly 1`] = `
<div>
  <div
    class="card-columns"
  >
    <div
      class="card"
      style="width: 18rem;"
    >
      <img
        class="card-img-top"
        src="image.jpg"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          sunt aut facere repellat provident occaecati excepturi optio reprehenderit
        </div>
        <p
          class="card-text"
        >
          quia et suscipitsuscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
        </p>
        <button
          bg="dark"
          class="btn btn-dark"
          type="button"
        >
          Read More
        </button>
      </div>
    </div>
    <div
      class="card"
      style="width: 18rem;"
    >
      <img
        class="card-img-top"
        src="image.jpg"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          qui est esse
        </div>
        <p
          class="card-text"
        >
          est rerum tempore vitaesequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
        </p>
        <button
          bg="dark"
          class="btn btn-dark"
          type="button"
        >
          Read More
        </button>
      </div>
    </div>
  </div>
</div>
`;
